docs/src/style.md

```diff
@@ -1,15 +1,78 @@
 Style guide and design principles
 =================================
 
+This document outlines JuMP's style guide. It also contains a set or principles
+for users writing JuMP models.
+
 Style guide
 -----------
 
-TODO: A style guide for JuMP, JuMP models and surrounding Julia code.
-Formatting, naming, use of macros, comments, TODOs, docstrings, etc.
+!!! info
+    For developers: if you modify JuMP's source code, please fix the style of
+    the surrounding code while you are there.
+
+### JuMP Source Code
+
+ - 80 character line limit. (Except for long URLs in comments.)
+ - 4 spaces for indentation.
+ - Use `lowercase_with_underscore` for variable and function names.
+ - Types and Modules should be in `CamelCase`.
+ - Use `UPPERCASE` for constants.
+ - No bangs (!) in function names. Use docstrings to indicate what is modified
+and variable lifetime issues.
+ - No one-letter variable names. This includes using `m = Model()`. The only
+exceptions are for indices in loops such as `for i in 1:N`.
+ - Start internal functions (i.e., those not meant to be used by a user) with an
+ underscore, e.g., `_add_constraint`.
+ - Use three backticks for name of internal function docstrings.
+ - Variables should never start with an underscore.
+ - To instantiate a vector, use `T[]` instead of `Vector{T}()`.
+ - Use proper punctuation and full sentences in comments.
+ - Use `for i in 1:N` instead of `for i = 1:N`.
+ - Never use trailing periods for floating point, e.g., `0.`.
+ - Never more than one consecutive blank line inside a function.
+ - Always use return from a function unless it is a one-line function.
+ - Do not split a one-line function over lines. Only if less than 80 characters.
+ - Type arguments to user-facing functions so that users do not get internal
+errors when passing wrong types.
+ - Imports: use `import` instead of `using`. Importing specific things from a
+ module is fine, stdlib imports first, alphabetical order after that.
```

Comment:
`using Module: foo` is strictly safer than `import Module: foo`, as it doesn't allow `Module.foo` to be overloaded. Do you really want to prohibit all usages of `using Module: foo`?